### ISSUE TITLE
Creating new firewall rule with azurerm_firewall_policy_rule_collection_group force existing rule to be recreated.

### ISSUE THREAD

**Issue by henryxuteck** [hashicorp/terraform-provider-azurerm#18114]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.1.9

### AzureRM Provider Version

3.19.1

### Affected Resource(s)/Data Source(s)

azurerm_firewall_policy_rule_collection_group

### Terraform Configuration Files

```hcl
resource "azurerm_firewall_policy_rule_collection_group" "example" {
  name               = "example-fwpolicy-rcg"
  firewall_policy_id = azurerm_firewall_policy.example.id
  priority           = 500

  network_rule_collection {
    name     = “rule 1”
	…
  }

  network_rule_collection {
    name     = “rule 2”
	…
  }

 network_rule_collection { <- this is a new rule
    name     = “rule 3”
	…
  }
  
}
```

### Debug Output/Panic Output

```shell
No matter whether the new rule is added after/between/before the previous rules, the plan always replaces rule2 with rule3 and recreates rule 2. This happens in `azurerm 3.0` only.
```

### Expected Behaviour

rule 3 should be added without impacting existing rules.

[This](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/guides/3.0-upgrade-guide#:~:text=other%20Terraform%20Providers.-,Firewall,-%3A%20The%20behavior) can be the cause, but I attempted to find a way to insert the rule3 to the end of the list by what I mentioned earlier, but I couldnt make it work.

**Comment by wuxu92** [comment]
Hi @henryxuteck just a simple confirmation: is azurerm 2.x works as your expect and from 3.x azurerm raised the recreate action?

**Comment by henryxuteck** [comment]
> Hi @henryxuteck just a simple confirmation: is azurerm 2.x works as your expect and from 3.x azurerm raised the recreate action?

Yes, 2.0 works fine, 3.0 doesn't.

**Comment by wuxu92** [comment]
Hi @juicybaba, I have tried your configuration by adding rule3 after rule2 and don't see the recreation of rule2. or how can we tell it is `the plan always replaces rule2 with rule3 and recreates rule 2.` I have reviewed the result in Azure Portal and the rules' order is the same as those listed in terraform.

In the update of azurerm 3.0, the Firewall changes from Set to List should not cause this kind of behavior

**Comment by juicybaba** [comment]
@wuxu92 thanks for getting back to me, i did a deeper look, and I wasn't able to reproduce the issue with the dummy rule collection name, but I am able to see the issue with the below collection name. I have my code and screenshot attached.

I would also need some help on understand the `order matters`, and what change is required.
```
The behavior of the nested items will be changed to List instead of Sets where required, meaning that the order of these items now matters. Note that if you're referencing these nested items within your Terraform Configuration, then this may require some code changes.
```

`azurerm_firewall_policy_rule_collection_group` resource
```
 resource "azurerm_firewall_policy_rule_collection_group" "default" {
  for_each           = local.firewall_rule_collection_groups
  name               = format("fwrcg-%s", each.key)
  firewall_policy_id = azurerm_firewall_policy.firewall_test.id
  priority           = each.value.priority

  dynamic "network_rule_collection" {
    iterator = self
    for_each = each.value.network_rule_collections
    content {
      name     = format("nrc-%s", self.key)
      priority = self.value.priority
      action   = self.value.action
      dynamic "rule" {
        for_each = self.value.rules
        content {
          name              = rule.key
          protocols         = rule.value.protocols
          destination_ports = rule.value.destination_ports

          destination_addresses = rule.value.destination_addresses
          destination_ip_groups = rule.value.destination_ip_groups
          destination_fqdns     = rule.value.destination_fqdns
          
          source_addresses = rule.value.source_addresses
          source_ip_groups = rule.value.source_ip_groups
        }
      }
    }
  }

  dynamic "application_rule_collection" {
    iterator = self
    for_each = each.value.application_rule_collections

    content {
      name     = format("arc-%s", self.key)
      priority = self.value.priority
      action   = self.value.action
      dynamic "rule" {
        for_each = lookup(self.value, "rules", {})
        content {
          name = rule.key
          
          destination_fqdns     = rule.value.destination_fqdns
          destination_fqdn_tags = rule.value.destination_fqdn_tags
          
          source_addresses = rule.value.source_addresses
          source_ip_groups = rule.value.source_ip_groups

          terminate_tls = false
          dynamic "protocols" {
            for_each = rule.value.protocols
            content {
              type = protocols.value.type
              port = protocols.value.port
            }
          }
        }
      }
    }
  }

  dynamic "nat_rule_collection" {
    iterator = self
    for_each = each.value.nat_rule_collections
    content {
      name     = format("natrc-%s", self.key)
      priority = self.value.priority
      action   = "Dnat"
      dynamic "rule" {
        for_each = lookup(self.value, "rules", {})
        content {
          name      = rule.key
          protocols = rule.value.protocols
          
          destination_address = rule.value.destination_address
          destination_ports   = rule.value.destination_ports
          translated_address  = rule.value.translated_address
          translated_port     = rule.value.translated_port
          
          source_addresses = rule.value.source_addresses
          source_ip_groups = rule.value.source_ip_groups
        }
      }
    }
  }

}
```


`firewall_rule_collection_groups` value: before adding new network rule collection.
```
locals {
  firewall_rule_collection_groups = {
    landingzones = {
      priority = 40000
      network_rule_collections = {
        Allow-TeckAzurePrivateSpokeObkCr-to-GoogleCloud-PostgreSql = {
          priority = 30000
          action   = "Allow"
          rules = {
            rule1 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["10.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["10.0.0.0/8"]
              source_ip_groups      = null
            }
            rule2 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["11.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["11.0.0.0/8"]
              source_ip_groups      = null
            }

          }
        },
        Allow-TeckAzurePrivateSpokeObkCr-to-TeckAzurePrivateSpokeObkData = {
          priority = 40000
          action   = "Allow"
          rules = {
            rule1 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["10.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["10.0.0.0/8"]
              source_ip_groups      = null
            }

            rule2 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["11.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["11.0.0.0/8"]
              source_ip_groups      = null
            }

          }
        }

      }
      application_rule_collections = {}
      nat_rule_collections         = {}
    }
  }
}
```

after: add `Allow-TeckAzurePrivateSpokeObkCr-to-Internet` network rule collection
```
locals {
  firewall_rule_collection_groups = {
    landingzones = {
      priority = 40000
      network_rule_collections = {
        Allow-TeckAzurePrivateSpokeObkCr-to-GoogleCloud-PostgreSql = {
          priority = 30000
          action   = "Allow"
          rules = {
            rule1 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["10.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["10.0.0.0/8"]
              source_ip_groups      = null
            }
            rule2 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["11.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["11.0.0.0/8"]
              source_ip_groups      = null
            }

          }
        },
        Allow-TeckAzurePrivateSpokeObkCr-to-TeckAzurePrivateSpokeObkData = {
          priority = 40000
          action   = "Allow"
          rules = {
            rule1 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["10.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["10.0.0.0/8"]
              source_ip_groups      = null
            }

            rule2 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["11.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["11.0.0.0/8"]
              source_ip_groups      = null
            }

          }
        },
        Allow-TeckAzurePrivateSpokeObkCr-to-Internet = {
          priority = 40000
          action   = "Allow"
          rules = {
            rule1 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["10.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["10.0.0.0/8"]
              source_ip_groups      = null
            }

            rule2 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["12.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["12.0.0.0/8"]
              source_ip_groups      = null
            }

          }
        }

      }
      application_rule_collections = {}
      nat_rule_collections         = {}
    }
  }
}
```

terraform plan: 
![image](https://user-images.githubusercontent.com/79539850/187567434-24a27c6c-f068-4f50-969b-2e06ac889358.png)

**Comment by juicybaba** [comment]
@wuxu92 
I would also need some help on understand the `order matters` part of below message, and what change is required.

Firewall: The behavior of the nested items will be changed to List instead of Sets where required, meaning that the `order of these items now matters`. Note that if you're referencing these nested items within your Terraform Configuration, then this may require some code changes.

**Comment by wuxu92** [comment]
I have tried to add `xxx2` after `xxx3` and the result order is listed as configuration and no recreation happened.

![image](https://user-images.githubusercontent.com/2633022/189295144-1c9b0fd4-bcdd-49fd-aadd-7fb043a1325e.png)

could you please share your full `azurerm_firewall_policy_rule_collection_group` configuration before and after applying to see what is the difference?

**Comment by wuxu92** [comment]
here is the origin PR of the change from `set` to `list`, and it looks fine to me. https://github.com/hashicorp/terraform-provider-azurerm/pull/15008

**Comment by juicybaba** [comment]
@wuxu92 thanks for getting back to me. Please see the full configuration for `azurerm_firewall_policy`, `azurerm_firewall_policy_rule_collection_group` and `local.firewall_rule_collection_groups`. The only difference between before and after is, a new network rule collection `network_rule_collections_3` get added to `local.firewall_rule_collection_groups`

Before:

```
resource "azurerm_firewall_policy" "firewall_test" {
  resource_group_name = module.resourcegroup.resource_group.name
  location            = module.resourcegroup.resource_group.location
  base_policy_id      = null
  name                = format("fwpol-test")
  # https://github.com/hashicorp/terraform-provider-azurerm/issues/9620
  tags                     = tomap({ for key, val in module.resourcegroup.resource_group.tags : lower(key) => val })
  sku                      = "Standard"
  threat_intelligence_mode = "Deny"
  dns { proxy_enabled = true }
}

resource "azurerm_firewall_policy_rule_collection_group" "default" {
  for_each           = local.firewall_rule_collection_groups
  name               = format("fwrcg-%s", each.key)
  firewall_policy_id = azurerm_firewall_policy.firewall_test.id
  priority           = each.value.priority

  dynamic "network_rule_collection" {
    iterator = self
    for_each = each.value.network_rule_collections
    content {
      name     = format("nrc-%s", self.key)
      priority = self.value.priority
      action   = self.value.action
      dynamic "rule" {
        for_each = self.value.rules
        content {
          name              = rule.key
          protocols         = rule.value.protocols
          destination_ports = rule.value.destination_ports

          destination_addresses = rule.value.destination_addresses
          destination_ip_groups = rule.value.destination_ip_groups
          destination_fqdns     = rule.value.destination_fqdns

          source_addresses = rule.value.source_addresses
          source_ip_groups = rule.value.source_ip_groups
        }
      }
    }
  }

  dynamic "application_rule_collection" {
    iterator = self
    for_each = each.value.application_rule_collections

    content {
      name     = format("arc-%s", self.key)
      priority = self.value.priority
      action   = self.value.action
      dynamic "rule" {
        for_each = lookup(self.value, "rules", {})
        content {
          name = rule.key

          destination_fqdns     = rule.value.destination_fqdns
          destination_fqdn_tags = rule.value.destination_fqdn_tags

          source_addresses = rule.value.source_addresses
          source_ip_groups = rule.value.source_ip_groups

          terminate_tls = false
          dynamic "protocols" {
            for_each = rule.value.protocols
            content {
              type = protocols.value.type
              port = protocols.value.port
            }
          }
        }
      }
    }
  }

  dynamic "nat_rule_collection" {
    iterator = self
    for_each = each.value.nat_rule_collections
    content {
      name     = format("natrc-%s", self.key)
      priority = self.value.priority
      action   = "Dnat"
      dynamic "rule" {
        for_each = lookup(self.value, "rules", {})
        content {
          name      = rule.key
          protocols = rule.value.protocols

          destination_address = rule.value.destination_address
          destination_ports   = rule.value.destination_ports
          translated_address  = rule.value.translated_address
          translated_port     = rule.value.translated_port

          source_addresses = rule.value.source_addresses
          source_ip_groups = rule.value.source_ip_groups
        }
      }
    }
  }

}

locals {
  firewall_rule_collection_groups = {
    landingzones = {
      priority = 40000
      network_rule_collections = {
        network_rule_collections_1 = {
          priority = 30000
          action   = "Allow"
          rules = {
            rule1 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["10.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["10.0.0.0/8"]
              source_ip_groups      = null
            }
            rule2 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["11.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["11.0.0.0/8"]
              source_ip_groups      = null
            }

          }
        },
        network_rule_collections_3 = {
          priority = 40000
          action   = "Allow"
          rules = {
            rule1 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["10.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["10.0.0.0/8"]
              source_ip_groups      = null
            }

            rule2 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["11.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["11.0.0.0/8"]
              source_ip_groups      = null
            }

          }
        }

      }
      application_rule_collections = {}
      nat_rule_collections         = {}
    }
  }
}
```
After:

```
resource "azurerm_firewall_policy" "firewall_test" {
  resource_group_name = module.resourcegroup.resource_group.name
  location            = module.resourcegroup.resource_group.location
  base_policy_id      = null
  name                = format("fwpol-test")
  # https://github.com/hashicorp/terraform-provider-azurerm/issues/9620
  tags                     = tomap({ for key, val in module.resourcegroup.resource_group.tags : lower(key) => val })
  sku                      = "Standard"
  threat_intelligence_mode = "Deny"
  dns { proxy_enabled = true }
}

resource "azurerm_firewall_policy_rule_collection_group" "default" {
  for_each           = local.firewall_rule_collection_groups
  name               = format("fwrcg-%s", each.key)
  firewall_policy_id = azurerm_firewall_policy.firewall_test.id
  priority           = each.value.priority

  dynamic "network_rule_collection" {
    iterator = self
    for_each = each.value.network_rule_collections
    content {
      name     = format("nrc-%s", self.key)
      priority = self.value.priority
      action   = self.value.action
      dynamic "rule" {
        for_each = self.value.rules
        content {
          name              = rule.key
          protocols         = rule.value.protocols
          destination_ports = rule.value.destination_ports

          destination_addresses = rule.value.destination_addresses
          destination_ip_groups = rule.value.destination_ip_groups
          destination_fqdns     = rule.value.destination_fqdns

          source_addresses = rule.value.source_addresses
          source_ip_groups = rule.value.source_ip_groups
        }
      }
    }
  }

  dynamic "application_rule_collection" {
    iterator = self
    for_each = each.value.application_rule_collections

    content {
      name     = format("arc-%s", self.key)
      priority = self.value.priority
      action   = self.value.action
      dynamic "rule" {
        for_each = lookup(self.value, "rules", {})
        content {
          name = rule.key

          destination_fqdns     = rule.value.destination_fqdns
          destination_fqdn_tags = rule.value.destination_fqdn_tags

          source_addresses = rule.value.source_addresses
          source_ip_groups = rule.value.source_ip_groups

          terminate_tls = false
          dynamic "protocols" {
            for_each = rule.value.protocols
            content {
              type = protocols.value.type
              port = protocols.value.port
            }
          }
        }
      }
    }
  }

  dynamic "nat_rule_collection" {
    iterator = self
    for_each = each.value.nat_rule_collections
    content {
      name     = format("natrc-%s", self.key)
      priority = self.value.priority
      action   = "Dnat"
      dynamic "rule" {
        for_each = lookup(self.value, "rules", {})
        content {
          name      = rule.key
          protocols = rule.value.protocols

          destination_address = rule.value.destination_address
          destination_ports   = rule.value.destination_ports
          translated_address  = rule.value.translated_address
          translated_port     = rule.value.translated_port

          source_addresses = rule.value.source_addresses
          source_ip_groups = rule.value.source_ip_groups
        }
      }
    }
  }

}

locals {
  firewall_rule_collection_groups = {
    landingzones = {
      priority = 40000
      network_rule_collections = {
        network_rule_collections_1 = {
          priority = 30000
          action   = "Allow"
          rules = {
            rule1 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["10.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["10.0.0.0/8"]
              source_ip_groups      = null
            }
            rule2 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["11.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["11.0.0.0/8"]
              source_ip_groups      = null
            }

          }
        },
        network_rule_collections_3 = {
          priority = 40000
          action   = "Allow"
          rules = {
            rule1 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["10.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["10.0.0.0/8"]
              source_ip_groups      = null
            }

            rule2 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["11.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["11.0.0.0/8"]
              source_ip_groups      = null
            }

          }
        },
        network_rule_collections_2 = {
          priority = 40000
          action   = "Allow"
          rules = {
            rule1 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["10.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["10.0.0.0/8"]
              source_ip_groups      = null
            }

            rule2 = {
              protocols             = ["TCP"]
              destination_ports     = ["*"]
              destination_addresses = ["12.0.0.0/8"]
              destination_ip_groups = null
              destination_fqdns     = null
              source_addresses      = ["12.0.0.0/8"]
              source_ip_groups      = null
            }

          }
        }

      }
      application_rule_collections = {}
      nat_rule_collections         = {}
    }
  }
}
```

**Comment by jr8279** [comment]
Has anyone found a solution for this issue? 
We are upgrading to a premium firewall, which requires us to move to Azure Firewall Policy and Firewall Rule Collection Groups. Currently with classic rules each rule collection gets its own tf file and allows for easy management of our rules at scale. I've attempted a solution similar to [juicybaba](https://github.com/juicybaba) with identical results; yamldecode orders the objects lexicographically by key and it forces all of our rule collections to be rewritten almost every time a new rule collection is added. In looking at the tfstate file, it appears that the Rule Collection Group gets an Azure Resource ID, but the rule collections do not; they are just an ordered list of objects. I'm stumped as to how our operations team can manage our 40+ rule collections without them all being in a single variable map and only adding new rule collections to the bottom.

**Comment by rgb000** [comment]
Save issue with terraform 1.3.0, provider version: 3.62.1

**Comment by jr8279** [comment]
There are several issues open with Hashicorp for the azurerm_firewall_policy_rule_collection_group module. As I've worked on this issue and talked to Microsoft reps about it I think the issue is with how Microsoft designed Firewall Policy Rule Collection Groups and not an issue with how Terraform deploys them.

[https://github.com/hashicorp/terraform-provider-azurerm/issues/16586](url)

Microsoft made the decision with the Rule Collection Groups to make the list of collections and subsequent rules an ordered list rather than making each collection a resource. This means that there isn't a way for Azure to identify each collection and thus if the list changes it must recreate the list. Also notice in Azure Portal that the Rule Collections in a Rule Collection Group are ordered in reverse order of which they are processed (Application, then Network, then NAT).

![image](https://github.com/hashicorp/terraform-provider-azurerm/assets/10197393/66a3c7f2-4220-4e73-ad78-09a9367d743d)


To reduce the potential impact to our deployments we broke the collections into individual yml files, and when we pull the collections together we sort them by priority. There will still be periodic changes to the list as we clean up old rules, but it is a higher probability that new rules will be placed at the bottom of the priority list. 

Microsoft has also confirmed with me that when updates are made to the firewall rules, the instance of the firewall will be removed from the pool, have the updates made to it, and then the instance will be returned to the pool before moving to the next firewall instance (a 1-2 minute process per instance). So any restructuring of the collections and rules will not impact production.

**Comment by ozen10** [comment]
> Microsoft has also confirmed with me that when updates are made to the firewall rules, the instance of the firewall will be removed from the pool, have the updates made to it, and then the instance will be returned to the pool before moving to the next firewall instance (a 1-2 minute process per instance). So any restructuring of the collections and rules will not impact production.

Hi @jr8279 just checking if I got this correctly, There should be no downtime on the network connectivity even if the collections groups and rules are being recreated?

**Comment by jr8279** [comment]
According to Microsoft, there shouldn’t be a downtime. But depending on how many firewalls instances you are using, there could be a period of time where some traffic is using the new rule list and some traffic is using the old rule list. They told me it takes 1-2 minutes per instance to update.

**Comment by dkooll** [comment]
rules are ok, only thing is if you remove a collection it still wants to modify existing ones... changing rule orders etc

**Comment by dkooll** [comment]
@wuxu92 

works great, only small thing is that there's still priority shift of the collections if you remove a collection that has a lower priority then the others

**Comment by LeonardoFronza** [comment]
Hey everyone.
Some get the solution for this problem?

I want to know how to resolve.

**Comment by vitor-antoni** [comment]
Still fronting this problem. Currently using 4.74 of Azure RM Provider version.
